### PR TITLE
perf(api): allow to use an arena for return values

### PR DESCRIPTION
**Pull request by bfredl**
complement/alternative to #19877 . While the lua special case for buffer text is still worth it, for other API functions we should eliminate return value allocation without special casing lua-api at all.

if an API function takes an `Arena *arena` it is expected to use this for allocate return values. it is also admissible to use pointers to static memory or pointers owned elsewhere, like in the `nvim_buf_get_name` example. Any callsite of api-functions must consume returned memory immediately before invoking anything which might change editor state (and thus such pointers). Arenas are now also lazily initialized, so there are no xmalloc/xfree calls involved if the arena actually is not used.

### CODE REVIEW

File: src/nvim/api/buffer.c [hunk 1]
```diff
@@ -1021,7 +1021,7 @@ void nvim_buf_del_var(Buffer buffer, String name, Error *err)
 /// @param buffer     Buffer handle, or 0 for current buffer
 /// @param[out] err   Error details, if any
 /// @return Buffer name
-String nvim_buf_get_name(Buffer buffer, Error *err)
+String nvim_buf_get_name(Buffer buffer, Arena *arena, Error *err)
```
Comment (lewis6991) [inline src/nvim/api/buffer.c:1024]: Is the arena argument required here?

Comment (bfredl) [reply src/nvim/api/buffer.c:1024]: Yes, otherwise, the wrapper would have attempted to free the return value.

Comment (lewis6991) [reply src/nvim/api/buffer.c:1024]: Ah yes, sorry of course.

**Comment by lewis6991** [comment]
Very nice!

I guess to optimize `nvim_buf_get_lines` we will still need something like #19877, so functions can have access to `lstate`. That is unless you have another magical idea on how not to special case Lua.

**Comment by bfredl** [comment]
Let's still do #19877 for now, `nvim_buf_get_lines` is used more than frequently enough from lua to warrant its own micro-optimization.

**Pull request merged**